File: and-bible/app/src/main/java/net/bible/android/view/activity/download/DocumentDownloadItemAdapter.java
package net.bible.android.view.activity.download;

import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;

import net.bible.android.activity.R;
import net.bible.android.control.ControlFactory;
import net.bible.android.control.download.DownloadControl;
import net.bible.android.view.activity.base.ListActionModeHelper;
import net.bible.android.view.activity.download.progress.DocumentDownloadProgressCache;
import net.bible.service.common.CommonUtils;

import org.crosswire.common.progress.JobManager;
import org.crosswire.common.progress.WorkEvent;
import org.crosswire.common.progress.WorkListener;
import org.crosswire.jsword.book.Book;
import org.crosswire.jsword.book.basic.AbstractPassageBook;
import org.crosswire.jsword.versification.system.SystemKJV;

import java.util.List;

/**
 * nice example here: http://shri.blog.kraya.co.uk/2010/04/19/android-multi-line-select-list/
 * 
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's author.
 */
public class DocumentDownloadItemAdapter extends ArrayAdapter<Book> {

	private int resource;
	
	private DownloadControl downloadControl = ControlFactory.getInstance().getDownloadControl();

	private DocumentDownloadProgressCache documentDownloadProgressCache;

	private WorkListener progressUpdater;

	private final ListActionModeHelper.ActionModeActivity actionModeActivity;

	private static int ACTIVATED_COLOUR = CommonUtils.getResourceColor(R.color.list_item_activated);

	public DocumentDownloadItemAdapter(Context _context, int _resource, List<Book> _items, ListActionModeHelper.ActionModeActivity actionModeActivity) {
		super(_context, _resource, _items);
		resource = _resource;
		this.actionModeActivity = actionModeActivity;

		// Listen for Progress changes and update the ui
		documentDownloadProgressCache = new DocumentDownloadProgressCache();

		progressUpdater = new WorkListener() {
			@Override
			public void workProgressed(WorkEvent ev) {
				documentDownloadProgressCache.updateProgress(ev.getJob());
			}

			@Override
			public void workStateChanged(WorkEvent ev) {
			}
		};
	}

	public void startMonitoringDownloads() {
		JobManager.addWorkListener(progressUpdater);
	}

	public void stopMonitoringDownloads() {
		if (progressUpdater!=null) {
			JobManager.removeWorkListener(progressUpdater);
		}
	}

	@Override
	public @NonNull View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {

		Book document = getItem(position);

		// Pick up the TwoLineListItem defined in the xml file
		DocumentDownloadListItem view;
		if (convertView == null) {
			LayoutInflater inflater = (LayoutInflater) getContext().getSystemService(Context.LAYOUT_INFLATER_SERVICE);
			view = (DocumentDownloadListItem) inflater.inflate(resource, parent, false);
		} else {
			view = (DocumentDownloadListItem) convertView;
			//  Get previous convertView doc to uncache
			documentDownloadProgressCache.documentListItemReallocated(view);
		}

		// remember which item is being shown
		view.setDocument(document);
		documentDownloadProgressCache.documentListItemShown(document, view);

		view.updateControlState(downloadControl.getBookInstallStatus(document));

		// Set value for the first text field
		if (view.getText1() != null) {
			// eBible repo uses abbreviation for initials and initials now contains the repo name!!!
			// but helpfully JSword uses initials if abbreviation does not exist, as will be the case for all other repos.
			String initials = document.getAbbreviation();
			view.getText1().setText(initials);
		}

		// set value for the second text field
		if (view.getText2() != null) {
			String name = document.getName();
			if (document instanceof AbstractPassageBook) {
				final AbstractPassageBook bible = (AbstractPassageBook)document;
				// display v11n name if not KJV
				if (!SystemKJV.V11N_NAME.equals(bible.getVersification().getName())) {
					name += " ("+bible.getVersification().getName()+")";
				}
			}
			view.getText2().setText(name);
		}

		if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
			if (actionModeActivity.isItemChecked(position)) {
				view.setBackgroundColor(ACTIVATED_COLOUR);
			} else {
				view.setBackgroundColor(Color.TRANSPARENT);
			}
		}

		return view;
	}
}

File: app/src/main/java/net/bible/android/view/activity/download/DocumentDownloadItemAdapter.kt
/*
 * Copyright (c) 2020 Martin Denham, Tuomas Airaksinen and the And Bible contributors.
 *
 * This file is part of And Bible (http://github.com/AndBible/and-bible).
 *
 * And Bible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * And Bible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with And Bible.
 * If not, see http://www.gnu.org/licenses/.
 *
 */
package net.bible.android.view.activity.download

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import net.bible.android.activity.R
import net.bible.android.activity.databinding.DocumentListItemBinding
import net.bible.android.control.download.DownloadControl
import net.bible.android.view.activity.base.DocumentSelectionBase
import net.bible.android.view.activity.base.RecommendedDocuments
import net.bible.service.common.Ref
import org.crosswire.jsword.book.Book

/**
 * nice example here: http://shri.blog.kraya.co.uk/2010/04/19/android-multi-line-select-list/
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 */
class DocumentDownloadItemAdapter(
    val activity: DocumentSelectionBase,
    private val downloadControl: DownloadControl,
    private val recommendedDocuments: Ref<RecommendedDocuments>
) : ArrayAdapter<Book>(activity, R.layout.document_list_item, ArrayList<Book>())
{
    private lateinit var bindings: DocumentListItemBinding

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        val document = getItem(position)!!

        bindings = if (convertView == null) {
            val inflater = activity.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            DocumentListItemBinding.inflate(inflater, parent, false)
        } else {
            DocumentListItemBinding.bind(convertView)
        }

        val view = (convertView?: bindings.root) as DocumentListItem
        view.binding = bindings

        // remember which item is being shown
        view.document = document
        view.recommendedDocuments = recommendedDocuments.value

        view.setIcons(downloadScreen = true)
        bindings.undoButton.setOnClickListener {
            downloadControl.cancelDownload(document)
        }

        bindings.aboutButton.setOnClickListener {
            activity.handleAbout(listOf(document))
        }

        view.updateControlState(downloadControl.getDocumentStatus(document))

        // Set value for the first text field
        // eBible repo uses abbreviation for initials and initials now contains the repo name!!!
        // but helpfully JSword uses initials if abbreviation does not exist, as will be the case for all other repos.
        val initials = document.abbreviation
        bindings.documentAbbreviation.text = initials

        // set value for the second text field
        bindings.documentName.text = document.name
        return view
    }
}
